### PR TITLE
Prune old rows in user_ips tables.

### LINKED ISSUE #4352: Ability to specify the retention period for user_ips table
We should surely default to 6M or something of user_ips retention, for GDPR compliance etc

### PR DESCRIPTION
**Pull request by erikjohnston**
~~Based on #6089. [Diff](https://github.com/matrix-org/synapse/compare/erikj/cleanup_user_ips...erikj/cleanup_user_ips_2).~~

Commits are independently reviewable.

Note that I've added a `wrap_as_background_process` so that we don't have to have a bunch of inline functions just to wrap a function in `run_as_background_process`. If this feels icky then I'm happy to remove it.

Fixes #4352

### CODE REVIEW

File: synapse/metrics/background_process_metrics.py [hunk 1]
```diff
@@ -197,7 +199,15 @@ def run():
                 _background_processes.setdefault(desc, set()).add(proc)
 
             try:
-                yield func(*args, **kwargs)
+                # We ensureDeferred here to handle coroutines
+                result = func(*args, **kwargs)
+
+                # We need this check because ensureDeferred doesn't like when
+                # func doesn't return a Deferred or coroutine.
```
Comment (richvdh) [inline synapse/metrics/background_process_metrics.py]: riiight... but why are we calling `run_as_background_process` on a thing which doesn't return a coroutine or a deferred?

File: synapse/metrics/background_process_metrics.py [hunk 2]
```diff
@@ -197,7 +199,15 @@ def run():
                 _background_processes.setdefault(desc, set()).add(proc)
 
             try:
-                yield func(*args, **kwargs)
+                # We ensureDeferred here to handle coroutines
```
Comment (richvdh) [inline synapse/metrics/background_process_metrics.py]: I thought it was legit to `yield` on a coroutine?

File: synapse/metrics/background_process_metrics.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/metrics/background_process_metrics.py]: I get a `RuntimeWarning: coroutine 'ClientIpStore._prune_old_user_ips' was never awaited` if I don't do this

File: synapse/metrics/background_process_metrics.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/metrics/background_process_metrics.py]: I get a `RuntimeWarning: coroutine 'ClientIpStore._prune_old_user_ips' was never awaited` if I don't do this

File: synapse/metrics/background_process_metrics.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/metrics/background_process_metrics.py]: https://github.com/matrix-org/synapse/blob/master/synapse/util/caches/expiringcache.py#L75-L80 is an example where we are periodically running code that isn't async. I *think* its still legit for it to have a logcontext set in that case?

Comment (richvdh) [reply synapse/metrics/background_process_metrics.py]: oh, right. Yes, I guess it is.

File: synapse/metrics/background_process_metrics.py (hunk 2 above)
Comment (richvdh) [reply synapse/metrics/background_process_metrics.py]: that is... surprising. We certainly `yield` on async functions elsewhere.

**Comment by richvdh** [comment]
can you rebase this to remove #6089?

File: synapse/metrics/background_process_metrics.py (hunk 2 above)
Comment (richvdh) [reply synapse/metrics/background_process_metrics.py]: ... but only where we have an ensureDeferred at the top of the call stack.

**Pull request merged**